cpp/core/shuffle/ShuffleWriter.h

```diff
@@ -23,201 +23,101 @@
 
 #include "memory/ArrowMemoryPool.h"
 #include "memory/ColumnarBatch.h"
-#include "memory/Evictable.h"
+#include "memory/Reclaimable.h"
 #include "shuffle/Options.h"
+#include "shuffle/PartitionWriter.h"
 #include "shuffle/Partitioner.h"
 #include "shuffle/Partitioning.h"
+#include "shuffle/ShuffleMemoryPool.h"
 #include "utils/Compression.h"
 
 namespace gluten {
 
-class ShuffleMemoryPool : public arrow::MemoryPool {
- public:
-  ShuffleMemoryPool(arrow::MemoryPool* pool) : pool_(pool) {}
-
-  arrow::MemoryPool* delegated() {
-    return pool_;
-  }
-
-  arrow::Status Allocate(int64_t size, int64_t alignment, uint8_t** out) override {
-    auto status = pool_->Allocate(size, alignment, out);
-    if (status.ok()) {
-      bytesAllocated_ += size;
-    }
-    return status;
-  }
-
-  arrow::Status Reallocate(int64_t old_size, int64_t new_size, int64_t alignment, uint8_t** ptr) override {
-    auto status = pool_->Reallocate(old_size, new_size, alignment, ptr);
-    if (status.ok()) {
-      bytesAllocated_ += (new_size - old_size);
-    }
-    return status;
-  }
-
-  void Free(uint8_t* buffer, int64_t size, int64_t alignment) override {
-    pool_->Free(buffer, size, alignment);
-    bytesAllocated_ -= size;
-  }
-
-  int64_t bytes_allocated() const override {
-    return bytesAllocated_;
-  }
-
-  int64_t max_memory() const override {
-    return pool_->max_memory();
-  }
-
-  std::string backend_name() const override {
-    return pool_->backend_name();
-  }
-
-  int64_t total_bytes_allocated() const override {
-    return pool_->total_bytes_allocated();
-  }
-
-  int64_t num_allocations() const override {
-    throw pool_->num_allocations();
-  }
-
- private:
-  arrow::MemoryPool* pool_;
-  uint64_t bytesAllocated_ = 0;
-};
-
-class ShuffleWriter : public Evictable {
+class ShuffleWriter : public Reclaimable {
  public:
   static constexpr int64_t kMinMemLimit = 128LL * 1024 * 1024;
 
   virtual arrow::Status split(std::shared_ptr<ColumnarBatch> cb, int64_t memLimit) = 0;
 
-  virtual arrow::Result<std::unique_ptr<arrow::ipc::IpcPayload>> createPayloadFromBuffer(
-      uint32_t partitionId,
-      bool reuseBuffers) = 0;
-
-  virtual arrow::Status evictPayload(uint32_t partitionId, std::unique_ptr<arrow::ipc::IpcPayload> payload) = 0;
+  virtual arrow::Status evictPartitionBuffers(uint32_t partitionId, bool reuseBuffers) = 0;
```

Comment:
Is the member function being used only from `cpp/velox`? If so should we push it down to `VeloxShuffleWriter`?

Reply:
the work should benefit remote shuffle as well

Reply:
Still I assume moving the function may not impact rss. Since rss is now implementing `PartitionWriter` API which doesn't have direct access to `ShuffleWriter` API.

@marin-ma Would you confirm? Thanks.

It's non-blocking as well so please proceed if needed.

Reply:
@zhztheplayer Yes. We can remove it. Rss uses VeloxShuffleWriter + CelebornPartitionWriter.